active-rfcs/0000-router-navigation-failures.md

```diff
@@ -0,0 +1,173 @@
+      // ...
+    }
+  }
+})
+
+// using async/await
+let failure = await router.push('/')
+if (failure) {
+  // ...
+}
+```
+
+By not rejecting the `Promise` when the navigation fails, we are avoiding _Uncaught (in promise)_ errors while still keeping the possibility to check if the Navigation failed or not.
+
+### Navigation Failures
+
+There are a few different navigation failures, to be able to react differently in your code
+
+Navigation failures can be differentiated through a `type` property. All possible values are hold by an `enum`, `NavigationFailureType`:
+
+- `cancelled`: `next(false)` inside of a navigation guard or a newer navigation took place while another one was ongoing.
+- `duplicated`: Navigating to the same location as the current one will cancel the navigation and not invoke any Navigation guard.
+
+On top of the `type` property, navigation failures also expose `from` and `to` properties, exactly like `router.afterEach`
+
+### Redirections
+
+Redirecting inside of a navigation guard with `next('/url')` is not a navigation failure by itself, as the navigation still takes place and ends up somewhere. To detect a navigation, specially during SSR, there is a `redirectedFrom` property accessible on the `currentRoute`.
+
+E.g.: imagine a navigation guard that redirects to `/login` when the user isn't authenticated:
+
+```js
+router.beforeEach((to, from, next) => {
+  // redirect to the login page if the target location requires authentication
+  if (to.meta.requiresAuth && !isAuthenticated) next('/login')
+  else next()
+})
+```
+
+When navigating to a location that requires authentication, we can retrieve the original location the user was trying to access via `redirectedFrom`:
+
+```js
+// user is not authenticated
+await router.push('/profile/dashboard')
+
+// `redirectedFrom` is a RouteLocationNormalized, like `currentRoute` but we are omitting
+// most properties to make the example readable
+router.currentRoute // { path: '/login', redirectedFrom: { path: '/profile/dashboard' } }
+```
+
+## Changes to `router.afterEach`
+
+Since `router.afterEach` also triggers when a navigation fails, we need a way to know if the navigation succeeded or failed. To do that, we introduce an extra parameter that contains the same _failure_ we could find in a resolved navigation:
+
+```js
+import { NavigationFailureType } from 'vue-router'
+
+router.afterEach((to, from, failure) => {
+  if (failure) {
+    if (failure.type === NavigationFailureType.canceled) {
+      // ...
+    }
+  }
+})
+```
+
+# Drawbacks
+
+- Breaking change although migration is relatively simple and in many cases will allow the developer to remove existing code
+
+# Alternatives
+
+- Differentiating `next(false)` from navigations that get overriden by more recent navigations by defining another Navigation Failure
+
+# Adoption strategy
+
+- `afterEach` and `onError` are relatively simple to migrate, most of the time they are not used many times either.
+- `router.push
```

Comment:
the rest of this line seems missing. is there some special consideration for migrating `router.push`?

Reply:
Thanks for noticing!